antlr_ast/ast.py

```diff
@@ -283,3 +175,560 @@ def reportContextSensitivity(
+
+            # _fields_spec can contain field multiple times
+            # e.g. x=a and x=b
+            if field_dict.get(name):
+                # or / elif behaviour
+                continue
+
+            # get node -----
+            field_dict[name] = cls.get_path(node, path)
+        return cls(node, field_dict)
+
+    @classmethod
+    def get_path(cls, node, path):
+        # TODO: can be defined on FieldNode too
+        result = node
+        for i in range(len(path)):
+            result = getattr(result, path[i], None)
+            if result is None:
+                break
+
+        return result
+
+    @classmethod
+    def bind_to_transformer(cls, transformer_cls, default_transform_method="from_spec"):
+        for rule in cls._rules:
+            if isinstance(rule, str):
+                cls_method = default_transform_method
+            else:
+                rule, cls_method = rule[:2]
+            transformer_method = cls.get_transformer(cls_method)
+            bind_to_transformer(transformer_cls, rule, transformer_method)
+
+    @classmethod
+    def get_transformer(cls, method_name):
+        """Get method to bind to visitor"""
+        transform_function = getattr(cls, method_name)
+        assert callable(transform_function)
+
+        def transformer_method(self, node):
+            kwargs = {}
+            if inspect.signature(transform_function).parameters.get("helper"):
+                kwargs["helper"] = self.helper
+            return transform_function(node, **kwargs)
+
+        return transformer_method
+
+
+# TODO: test: if 'visit' in method, it has to be as 'visit_'
+class BaseNodeTransformer(NodeTransformer):
+    def __init__(self, registry: BaseNodeRegistry):
+        self.helper = TransformerHelper(registry)
+
+    def visit(self, node):
+        # TODO: I think transform_  + node.__class__.__name__ would be better/clearer then
+        #  as the node methods don't need to do any visiting (which is completely done by visit and generic_visit)
+        method = "visit_" + node.__class__.__name__
+        transformer = getattr(self, method, None)
+
+        if transformer is None:
+            return self.generic_visit(node)
+        else:
+            alias = transformer(node)
+            if isinstance(alias, AliasNode) or alias == node:
+                # this prevents infinite recursion and visiting
+                # AliasNodes with a name that is also the name of a BaseNode
+                self.generic_visit(alias)
+            else:
+                # visit BaseNode (e.g. result of Transformer method)
+                if isinstance(alias, list):
+                    # Transformer method can return array instead of node
+                    alias = [self.visit(el) for el in alias]  # TODO: test
+                elif isinstance(alias, BaseNode):
+                    alias = self.visit(alias)
+
+            return alias
+
+    def visit_Terminal(self, terminal):
```

Comment:
When will this or `visit_str` be reached? Are they just here because `Terminal` or `str` don't have `_fields`?

Reply:
These methods are called when visiting the tree leafs. Depending on whether debug mode is enabled, those will be strings or a Terminal (which is a string wrapper).